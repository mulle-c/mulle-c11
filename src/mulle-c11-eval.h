//
//  mulle_c11_eval.h
//
//  Copyright Â© 2022 Mulle kybernetiK. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_c11_eval_h__
#define mulle_c11_eval_h__

// MEMO: currently not exposed in mulle-c11.h, because it's fairly obscure
//       and rarely needed

/*
 * EVALUATION FUNCTIONS FOR PREPROCESSOR VARARGS
 */
//
// https://renenyffenegger.ch/notes/development/languages/C-C-plus-plus/preprocessor/macros/__VA_ARGS__/count-arguments
//

// treat __VA_ARGS__ like a C array, index starts at 0

// MULLE_C_VA_ARGS_0 can be called MULLE_C_VA_ARGS_0() as a0 will be nothing
#define MULLE_C_VA_ARGS_0(  a0, ...) a0
#define MULLE_C_VA_ARGS_1(  a0, a1, ...) a1
#define MULLE_C_VA_ARGS_2(  a0, a1, a2, ...) a2
#define MULLE_C_VA_ARGS_3(  a0, a1, a2, a3, ...) a3
#define MULLE_C_VA_ARGS_4(  a0, a1, a2, a3, a4, ...) a4
#define MULLE_C_VA_ARGS_5(  a0, a1, a2, a3, a4, a5, ...) a5
#define MULLE_C_VA_ARGS_6(  a0, a1, a2, a3, a4, a5, a6, ...) a6
#define MULLE_C_VA_ARGS_7(  a0, a1, a2, a3, a4, a5, a6, a7, ...) a7
#define MULLE_C_VA_ARGS_8(  a0, a1, a2, a3, a4, a5, a6, a7, a8, ...) a8
#define MULLE_C_VA_ARGS_9(  a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, ...) a9
#define MULLE_C_VA_ARGS_10( a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, ...) a10
#define MULLE_C_VA_ARGS_11( a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, ...) a11
#define MULLE_C_VA_ARGS_12( a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, ...) a12
#define MULLE_C_VA_ARGS_13( a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, ...) a13
#define MULLE_C_VA_ARGS_14( a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, ...) a14
#define MULLE_C_VA_ARGS_15( a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, ...) a15

#define MULLE_C_VA_ARGS_COUNT( ...) \
   MULLE_C_VA_ARGS_15( dummy __VA_OPT__(,) __VA_ARGS__, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0)

//
// https://github.com/pfultz2/Cloak/wiki/C-Preprocessor-tricks,-tips,-and-idioms
//
#define MULLE_C_IS_EMPTY( ...)  \
        _MULLE_C_IS_ ##  __VA_OPT__( NON_) ## EMPTY
#define _MULLE_C_IS_NON_EMPTY    0
#define _MULLE_C_IS_EMPTY        1

#define MULLE_C_VA_ARGS_WITH_DEFAULT( d, ...)  \
        _MULLE_C_ ##  __VA_OPT__( NON_) ## EMPTY_VA_ARGS_WITH_DEFAULT( d, __VA_ARGS__ )
#define MULLE_C_VA_ARGS_0_WITH_DEFAULT( d, ...)  \
        _MULLE_C_ ##  __VA_OPT__( NON_) ## EMPTY_VA_ARGS_WITH_DEFAULT( d, MULLE_C_VA_ARGS_0( __VA_ARGS__))

#define _MULLE_C_NON_EMPTY_VA_ARGS_WITH_DEFAULT( d, ...)  __VA_ARGS__
#define _MULLE_C_EMPTY_VA_ARGS_WITH_DEFAULT( d, ...)      d


#define MULLE_C_CAT( a, ...)             MULLE_C_PRIMITIVE_CAT( a, __VA_ARGS__)
#define MULLE_C_PRIMITIVE_CAT( a, ...)   a ## __VA_ARGS__

#define MULLE_C_IIF( c)          MULLE_C_PRIMITIVE_CAT( MULLE_C_IIF_, c)
#define MULLE_C_IIF_0( t, ...)   __VA_ARGS__
#define MULLE_C_IIF_1( t, ...)   t

#define MULLE_C_EAT(...)
#define MULLE_C_EXPAND(...)      __VA_ARGS__
#define MULLE_C_WHEN( c)         MULLE_C_IIF(c)( MULLE_C_EXPAND, MULLE_C_EAT)

#define MULLE_C_EMPTY()
#define MULLE_C_DEFER( id)       id MULLE_C_EMPTY()
#define MULLE_C_OBSTRUCT(...)    __VA_ARGS__ MULLE_C_DEFER( MULLE_C_EMPTY)()

//
// The "eval" trick can be done up to a loop depth of 256. If you need more,
// define more. The default is 64. The higher you go the more the compiler
// slows down though.
//
#define MULLE_C_EVAL256(...)   MULLE_C_EVAL128( MULLE_C_EVAL127( __VA_ARGS__))
#define MULLE_C_EVAL255(...)   MULLE_C_EVAL127( MULLE_C_EVAL127( __VA_ARGS__))
#define MULLE_C_EVAL128(...)   MULLE_C_EVAL64( MULLE_C_EVAL63( __VA_ARGS__))
#define MULLE_C_EVAL127(...)   MULLE_C_EVAL63( MULLE_C_EVAL63( __VA_ARGS__))
#define MULLE_C_EVAL64(...)    MULLE_C_EVAL32( MULLE_C_EVAL31( __VA_ARGS__))
#define MULLE_C_EVAL63(...)    MULLE_C_EVAL31( MULLE_C_EVAL31( __VA_ARGS__))
#define MULLE_C_EVAL32(...)    MULLE_C_EVAL16( MULLE_C_EVAL15( __VA_ARGS__))
#define MULLE_C_EVAL31(...)    MULLE_C_EVAL15( MULLE_C_EVAL15( __VA_ARGS__))
#define MULLE_C_EVAL16(...)    MULLE_C_EVAL8( MULLE_C_EVAL7( __VA_ARGS__))
#define MULLE_C_EVAL15(...)    MULLE_C_EVAL7( MULLE_C_EVAL7( __VA_ARGS__))
#define MULLE_C_EVAL8(...)     MULLE_C_EVAL4( MULLE_C_EVAL3( __VA_ARGS__))
#define MULLE_C_EVAL7(...)     MULLE_C_EVAL3( MULLE_C_EVAL3( __VA_ARGS__))
#define MULLE_C_EVAL4(...)     MULLE_C_EVAL2( MULLE_C_EVAL1( __VA_ARGS__))
#define MULLE_C_EVAL3(...)     MULLE_C_EVAL1( MULLE_C_EVAL1( __VA_ARGS__))
#define MULLE_C_EVAL2(...)     MULLE_C_EVAL1( __VA_ARGS__)
#define MULLE_C_EVAL1(...)     __VA_ARGS__

#define MULLE_C_EVAL(...)      MULLE_C_EVAL64( __VA_ARGS__)

#define MULLE_C_HAS_ARGS( ...)     MULLE_C_HAS_ARGS_ ## __VA_OPT__(1) ()
#define MULLE_C_HAS_ARGS_()        0
#define MULLE_C_HAS_ARGS_1()       1

// this is actually original, yay
#define MULLE_C_FOR_EACH_WITH_SEPARATOR( macro, s, sep, first, ...) \
    macro( first, s) __VA_OPT__( sep ) \
    MULLE_C_WHEN( MULLE_C_HAS_ARGS(__VA_ARGS__)) \
    ( \
        MULLE_C_OBSTRUCT( MULLE_C_FOR_EACH_WITH_SEPARATOR_INDIRECT) () \
        ( \
            macro, s ## i, sep, __VA_ARGS__ \
        ) \
    )

#define MULLE_C_FOR_EACH_WITH_SEPARATOR_INDIRECT() \
    MULLE_C_FOR_EACH_WITH_SEPARATOR

#define MULLE_C_FOR_EACH( macro, first, ...) \
    MULLE_C_FOR_EACH_WITH_SEPARATOR( macro, v, , first __VA_OPT__(,) __VA_ARGS__)


//
#define _MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR( macro, s, first, ...) \
    macro( first, s) __VA_OPT__( , ) \
    MULLE_C_WHEN( MULLE_C_HAS_ARGS(__VA_ARGS__)) \
    ( \
        MULLE_C_OBSTRUCT( _MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR_INDIRECT) () \
        ( \
            macro, s ## i, __VA_ARGS__ \
        ) \
    )

#define _MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR_INDIRECT()  \
   _MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR

// Emit a comma separated list. Macro should be defined like this:
//
// #define macro( argument, identifier)
//
// Argument will be each vararg in sequence. So with three varargs, the
// macro will be called three times.
// The identifier will be unique within the expansion (but very ugly).
// Call this by wrapping it in MULLE_C_EVAL
// e.g.
// #define field( value, identifier)   identifier = value
// MULLE_C_EVAL( MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR( field, 1, 2, 3))
// Expect: v = 1, vi = 2, vii = 3
//

#define MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR( macro, first, ...) \
    _MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR( macro, v, first __VA_OPT__(,) __VA_ARGS__)


// Example:
// Create a struct from values and initialize it.
// Step 1. Figure out type of value with __typeof__, create a field in the
//         struct for it (STRUCT_FIELD).
// Step 2. Create an initializer with the values (STRUCT_VALUE)
// Step 3. Create the struct (STRUCT_FOR_VALUES)
// Step 4. Execute (MULLE_C_EVAL)
//
// #define STRUCT_FIELD( expr, s) __typeof__( expr)  s;
// #define STRUCT_VALUE( expr, s) .s = expr
//
// #define STRUCT_FOR_VALUES( name, ...)                                                 \
//    struct                                                                             \
//    {                                                                                  \
//      MULLE_C_EVAL( MULLE_C_FOR_EACH( STRUCT_FIELD, __VA_ARGS__))                      \
//    } name =                                                                           \
//    {                                                                                  \
//      MULLE_C_EVAL( MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR( STRUCT_VALUE, __VA_ARGS__)) \
//    }
// MULLE_C_EVAL( STRUCT_FOR_VALUES( my_struct, 1.0f, "VfL Bochum", 1848LL))

#endif
