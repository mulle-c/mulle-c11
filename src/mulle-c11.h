//
//  mulle_c11.h
//
//  Copyright © 2016 Mulle kybernetiK. All rights reserved.
//  Copyright © 2016 Codeon GmbH. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//

#ifndef mulle_c11_h__
#define mulle_c11_h__

#ifdef __cplusplus
# error "C++ ? You're doing it wrong."
#endif


#ifndef __has_feature          // Optional.
# define __has_feature( x) 0   // Compatibility with non-clang compilers.
#endif

#ifndef __has_attribute
# define __has_attribute( x) 0
#endif


#define MULLE_C11_VERSION  ((4 << 20) | (1 << 8) | 0)


//
// basic C11 stuff, that's missing on "some" platforms
// https://gcc.gnu.org/wiki/C11Status
//
#ifndef MULLE_C11_NO_STDALIGN
# ifdef _WIN32
#  define MULLE_C11_NO_STDALIGN
# endif

# ifndef __clang__
#  ifdef __GNUC__
#   if __GNUC__ < 4 || (__GNUC__ == 4 && __GNUC_MINOR__ < 7)
#    define MULLE_C11_NO_STDALIGN
#   endif
#  endif
# endif
#endif


#ifdef MULLE_C11_NO_STDALIGN
# define alignof( x)    __alignof( x)
#else
# include <stdalign.h>
#endif


#ifdef _WIN32
# define __builtin_expect( x, y) x
#endif

// The following definitions are for the "outside looking in" not
// for compiling the (shared) library itself.
//
// https://msdn.microsoft.com/en-us/library/aa271769(v=vs.60).aspx
//
// If you are compiling a Windows DLL with constituent static libraries use
// -DMULLE_C_EXTERN_GLOBAL=extern. This ensures that global variables of those
// libraries are internally (within the DLL) linked.
//
// > Remember a DLL can't link with itself and the __imp_ stuff is generated
// > by the linker and is not available in the static LIBs
//
// Using -DMULLE_C_EXTERN_GLOBAL fails if your DLL links against other
// mulle_c11 derived dlls, because of conflicting settings. Therefore
// it's best to define and use your own package MULLE_C_EXTERN_GLOBAL
//
// e.g.
//    #ifndef MULLE_OBJC_EXTERN_GLOBAL
//      # define MULLE_OBJC_EXTERN_GLOBAL  MULLE_C_EXTERN_GLOBAL
//    #endif
// and override that with
// -DMULLE_OBJC_EXTERN_GLOBAL=extern
//
#ifdef _WIN32
# ifndef MULLE_C_EXTERN_GLOBAL
#  define MULLE_C_EXTERN_GLOBAL   extern  __declspec( dllimport)
# endif
# define MULLE_C_GLOBAL           __declspec( dllexport)
#else
# ifndef MULLE_C_EXTERN_GLOBAL
#  define MULLE_C_EXTERN_GLOBAL   extern
# endif
# define MULLE_C_GLOBAL
#endif

//
// stuff only clang and gcc understands
//
#if defined( __clang__) || defined( __GNUC__)

# define MULLE_C_ALWAYS_INLINE         __attribute__(( always_inline))
# define MULLE_C_NEVER_INLINE          __attribute__(( noinline))

# define MULLE_C_CONST_RETURN          __attribute__(( const))
# define MULLE_C_NO_RETURN             __attribute__(( noreturn))

# define MULLE_C_DEPRECATED            __attribute__(( deprecated))

// use these for function pointer modifiers (because win...)
# define _MULLE_C_NO_RETURN            MULLE_C_NO_RETURN
# define _MULLE_C_NEVER_INLINE         MULLE_C_NEVER_INLINE

#else

# ifdef _WIN32
#  define MULLE_C_NO_RETURN            __declspec( noreturn)
#  define MULLE_C_NEVER_INLINE         __declspec( noinline)

#  define _MULLE_C_NO_RETURN
#  define _MULLE_C_NEVER_INLINE

// check this https://stackoverflow.com/questions/1113409/attribute-constructor-equivalent-in-vc
// for MULLE_C_CONSTRUCTOR

# else
#  define MULLE_C_NO_RETURN
#  define MULLE_C_NEVER_INLINE
#  define _MULLE_C_NO_RETURN
#  define _MULLE_C_NEVER_INLINE
# endif

# define MULLE_C_ALWAYS_INLINE
# define MULLE_C_CONST_RETURN
# define MULLE_C_DEPRECATED

#endif

// clang only (?)
// gcc (at least < 4.8) doesn't like ‘returns_nonnull’ apparently ?

#if defined( __clang__)
# define MULLE_C_NONNULL_RETURN       __attribute__(( returns_nonnull))
#else
# define MULLE_C_NONNULL_RETURN
#endif


// some composites

#define MULLE_C_CONST_NONNULL_RETURN                 MULLE_C_NONNULL_RETURN \
                                                     MULLE_C_CONST_RETURN

#define MULLE_C_ALWAYS_INLINE_NONNULL_RETURN         MULLE_C_NONNULL_RETURN \
                                                     MULLE_C_ALWAYS_INLINE

#define MULLE_C_ALWAYS_INLINE_NONNULL_CONST_RETURN   MULLE_C_NONNULL_RETURN \
                                                     MULLE_C_ALWAYS_INLINE \
                                                     MULLE_C_CONST_RETURN


//
// cross platform __attribute__((constructor))
// place in front of function w/o ;
// https://stackoverflow.com/questions/1113409/attribute-constructor-equivalent-in-vc
// Use C++ initialized. C (.CRT$XIU) doesn't work, I guess, because on windows
// it returns a value (possibly terminating) and we are void here.
//
// The underscore method is global and will conflict with other symbols!
//

#ifdef _MSC_VER
# pragma section(".CRT$XCU",read)
# define MULLE_C_CONSTRUCTOR2_( f, p) \
        static void f(void); \
        __declspec(allocate(".CRT$XCU")) void (*f##_)(void) = f; \
        __pragma(comment(linker,"/include:" p #f "_"))
# ifdef _WIN64
#  define MULLE_C_CONSTRUCTOR( f) MULLE_C_CONSTRUCTOR2_(f,"")
# else
#  define MULLE_C_CONSTRUCTOR (f) MULLE_C_CONSTRUCTOR2_(f,"_")
# endif
#else
# define MULLE_C_CONSTRUCTOR( f) \
        __attribute__((constructor))
#endif


//
// cross platform __attribute__((destructor)). It's not really clear when and
// if this runs though. This is just compiler decl stuff.
//
// https://www.gonwan.com/2014/02/13/msvc-crt-initialization/
// ".CRT$XTU" is basically guessed
//
// Use XTU for "user" terminator C which is void:
// https://github.com/wyrover/book-code/blob/master/Linux%E4%B8%8B%E7%9A%84C%E5%BA%93%E5%87%BD%E6%95%B0%E6%BA%90%E4%BB%A3%E7%A0%81/%E6%9C%80%E5%85%A8%E7%9A%84Linux%E4%B8%8B%E7%9A%84C%E5%BA%93%E5%87%BD%E6%95%B0%E6%BA%90%E4%BB%A3%E7%A0%811/src/crt0dat.c
//
#ifdef _MSC_VER
# pragma section(".CRT$XTU",read)
# define MULLE_C_DESTRUCTOR2_( f, p) \
        static void f(void); \
        __declspec(allocate(".CRT$XTU")) void (*f##_)(void) = f; \
        __pragma(comment(linker,"/include:" p #f "_"))
# ifdef _WIN64
#  define MULLE_C_DESTRUCTOR( f) MULLE_C_DESTRUCTOR2_(f,"")
# else
#  define MULLE_C_DESTRUCTOR( f) MULLE_C_DESTRUCTOR2_(f,"_")
# endif
#else
# define MULLE_C_DESTRUCTOR( f) \
        __attribute__((destructor))
#endif



//
// Some warnings are just too much for regular builds.
// Sometimes if one has a weird error it's nice to enable
// this though. So it is recommeded to compile with
// -DMULLE_C11_NO_NOOB_WARNINGS=1 by default
//
// The alternative would be to gum up the code...
//
#if MULLE_C11_NO_NOOB_WARNINGS

# if defined( __clang__)
// if( x = next()) , don't need that warning all the time
#  pragma clang diagnostic ignored "-Wparentheses"

// switch( enum), don't need missing case warning ever really
#  pragma clang diagnostic ignored "-Wswitch"

// That's what a category is good for...
#  pragma clang diagnostic ignored "-Wobjc-protocol-method-implementation"
# endif


# if defined( __GNUC__)
#  pragma GCC diagnostic ignored "-Wparentheses"
#  pragma GCC diagnostic ignored "-Wswitch"
#  pragma GCC diagnostic ignored "-Wobjc-protocol-method-implementation"
# endif

#endif


//
// mark functions arguments a nonnull. This will indicate that a null argument
// will make the the function crash. Nothing more, nothing less.
//
// For mulle code adorn only `_` prefixed functions with MULLE_C_NONNULL_.
// Non-prefixed functions must not crash on null input, consequently must check
// input parameters and then the MULLE_C_NONNULL_ is a hindrance.
//
// Add to these defines as desired. Remember, more than five arguments:
// wrap them into a struct!
//
#if defined( __clang__) || defined( __GNUC__)
# define MULLE_C_NONNULL_FIRST                __attribute__((nonnull(1)))
# define MULLE_C_NONNULL_FIRST_SECOND         __attribute__((nonnull(1,2)))
# define MULLE_C_NONNULL_FIRST_SECOND_THIRD   __attribute__((nonnull(1,2,3)))
# define MULLE_C_NONNULL_FIRST_SECOND_FOURTH  __attribute__((nonnull(1,2,4)))
# define MULLE_C_NONNULL_FIRST_SECOND_FIFTH   __attribute__((nonnull(1,2,5)))
# define MULLE_C_NONNULL_FIRST_SECOND_FOURTH_FIFTH __attribute__((nonnull(1,2,4,5)))
# define MULLE_C_NONNULL_FIRST_THIRD          __attribute__((nonnull(1,3)))
# define MULLE_C_NONNULL_FIRST_THIRD_FOURTH   __attribute__((nonnull(3,4)))
# define MULLE_C_NONNULL_FIRST_FOURTH         __attribute__((nonnull(1,4)))
# define MULLE_C_NONNULL_FIRST_FIFTH          __attribute__((nonnull(1,5)))
# define MULLE_C_NONNULL_SECOND               __attribute__((nonnull(2)))
# define MULLE_C_NONNULL_SECOND_THIRD         __attribute__((nonnull(2,3)))
# define MULLE_C_NONNULL_THIRD                __attribute__((nonnull(3)))
# define MULLE_C_NONNULL_FOURTH               __attribute__((nonnull(4)))
# define MULLE_C_NONNULL_THIRD_FOURTH         __attribute__((nonnull(3,4)))
# define MULLE_C_NONNULL_FIFTH                __attribute__((nonnull(4)))
#else
# define MULLE_C_NONNULL_FIRST
# define MULLE_C_NONNULL_FIRST_SECOND
# define MULLE_C_NONNULL_FIRST_SECOND_THIRD
# define MULLE_C_NONNULL_FIRST_SECOND_FOURTH
# define MULLE_C_NONNULL_FIRST_SECOND_FIFTH
# define MULLE_C_NONNULL_FIRST_SECOND_FOURTH_FIFTH
# define MULLE_C_NONNULL_FIRST_THIRD
# define MULLE_C_NONNULL_FIRST_THIRD_FOURTH
# define MULLE_C_NONNULL_FIRST_FOURTH
# define MULLE_C_NONNULL_FIRST_FIFTH
# define MULLE_C_NONNULL_SECOND
# define MULLE_C_NONNULL_SECOND_THIRD
# define MULLE_C_NONNULL_THIRD
# define MULLE_C_NONNULL_THIRD_FOURTH
# define MULLE_C_NONNULL_FOURTH
# define MULLE_C_NONNULL_FIFTH
#endif

#endif // h
