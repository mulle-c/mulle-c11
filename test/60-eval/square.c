/* THIS TEST WAS WRITTEN BY PHIND. It missed initially
  1. SQUARE needs a dummy identifier parameter
  2. MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR needs to be wrapped in MULLE_C_EVAL
  3. <mulle-c11/mulle-c11-eval.h> needs to be included.
  But it came up, with the wrapper idea which is nice. Unfortunately it messed
  up the wrapper, but anyway...
*/

/*
The MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR macro is a convenient utility provided by the Mulle C11 library. It is designed to apply a given macro to a list of arguments, separating the results with commas. This can be useful when working with variadic arguments or other situations where you need to perform an operation on multiple items and separate the results with a comma.

Here's a breakdown of how the MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR macro works:

1. It takes three arguments: a macro, a list of arguments, and a final argument. The macro is applied to each element in the list, and the results are separated by commas.

2. The macro uses a series of other helper macros to handle the iteration and comma separation. These include MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR_1, MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR_2, MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR_3, and so on. Each of these helper macros takes care of applying the given macro to one or more elements in the list and separating the results with commas.

3. The MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR macro is implemented using variadic macro arguments, which allows it to accept a variable number of arguments for the list. This makes it highly flexible and adaptable to various use cases.

Here's an example of how you might use the MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR macro:
*/

#include <stdio.h>
#include <mulle-c11/mulle-c11.h>
#include <mulle-c11/mulle-c11-eval.h>

#define SQUARE(x) ((x) * (x))
#define WRAPPED_SQUARE(x, _) SQUARE(x)

int main() {
  int result[] = { MULLE_C_EVAL( MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR(WRAPPED_SQUARE, 1, 2, 3, 4) ) };

  for (size_t i = 0; i < sizeof(result) / sizeof(result[0]); ++i) {
    printf("result[%zu] = %d\n", i, result[i]);
  }

  return 0;
}


/*
In this example, the MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR macro is used to apply the SQUARE macro to a list of numbers (1, 2, 3, 4). The results are separated by commas and stored in an array. When compiled and executed, the output will be:

```
result[0] = 1
result[1] = 4
result[2] = 9
result[3] = 16
```

As you can see, the MULLE_C_FOR_EACH_WITH_COMMA_SEPARATOR macro provides a convenient way to perform operations on multiple items and separate the results with commas, making it a useful tool in the Mulle C11 library.
*/