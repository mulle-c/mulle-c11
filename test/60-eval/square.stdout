result[0] = 1
result[1] = 4
result[2] = 9
result[3] = 16
